cpp/include/typedb/common/exception.hpp

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (C) 2022 Vaticle
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#pragma once
+
+#include <stdexcept>
+#include <string>
+
+#include "typedb/common/native.hpp"
+
+namespace TypeDB {
+
+class TypeDBDriverException : public std::runtime_error {
+    const std::string errCode;
+    const std::string errMsg;
+
+   public:
+    TypeDBDriverException(const char* code, const char* message);
+    const std::string& code();
+    const std::string& message();
+    const char* what() const noexcept override;
+
+    static void check_and_throw();
+};
+
+typedef TypeDBDriverException TypeDBDriverCustomException;
```

Comment:
How should I handle errors which aren't due to TypeDB? Such as users accessing a moved object? TypeDBDriver but with errorCode INTERNAL?
I've currently used TypeDBDriverCustomException to keep them separate.